### DISCUSSION TITLE
Simple MQTT Client UI using this project

### DISCUSSION THREAD

**Discussion by sukesh-ak** [sukesh-ak/BSP-IDF5-ESP_LCD-LVGL9#6] [General]
Lets discuss the requirements for a Simple MQTT Client UI for ESP32-XX Display devices

@mhaberler

**Comment by mhaberler** [comment]
my goal would be to have generic, small footprint UI support which can can be dropped in as a library or component, not another tasmota or espeasy

let's assume the whole config can be mapped onto a JSON file which is either onboard in nvs or fetched on startup, which includes broker, credentials etc

looking at the broker->UI path I'd see:

1. list of topics to subscribe
2. which LVGL objects/widgets that maps to
3. some way of transmogrifying the incoming message into something which fits with the UI object

1+2 are easy
3:  so for example we get

> Topic: home/OMG_heltec_ble/BTtoMQTT/DD79C68FBDA2QoS: 0
> 
> {"id":"DD:79:C6:8F:BD:A2","mac_type":1,"adv_type":0,"name":"Ruuvi BDA2","manufacturerdata":"9904050858809dfffffec8fc28001c81b681030add79c68fbda2","rssi":-82,"brand":"Ruuvi","model":"RuuviTag","model_id":"RuuviTag_RAWv2","type":"ACEL","track":true,"tempc":10.68,"tempf":51.224,"hum":82.3125,"pres":1155.35,"accx":-0.305967,"accy":-0.964974,"accz":0.027459,"volt":2.637,"tx":4,"mov":129,"seq":778,"mac":"DD:79:C6:8F:BD:A2"}

assume I want to associate "tempc" with a UI label
I can parse the message and extract a key in C++ but thats inflexible which suggests an interpretive step in between

something like [JSONata expressions](https://docs.jsonata.org/expressions) would be cool but that's from nodejs land and not c++

embedding [berry](https://github.com/berry-lang/berry) would be an option

btw I'd try to keep those parts esp-unaware to the extent possible because they are generic

what do you think?

**Reply by sukesh-ak** [reply]
Sounds like a larger project in time required.
But luckily, someone has already thought about it and created similar project here
https://github.com/YooSHome/SenseCapConfig/

They have built it for square screen based on `Seeed Studio SenseCap Indicator` device.

**Reply by sukesh-ak** [reply]
Its using `lvgl 8x` so `lv_observer` option is not there.

**Reply by mhaberler** [reply]
looking at my previous post: too ambitious. Strike out the interpreter, too much hassle.
on C++ level, even a mapping table would be a fine start.

I like the JSON config approach in SenseCapConfig

It depends a bit on how the LVGL/Squareline folk go forward integrating observers into Squareline or whatever follows that

**Reply by sukesh-ak** [reply]
LVGL & SquareLine has parted ways. So I am looking forward to what comes next from LVGL side
https://blog.lvgl.io/2024-02-12/sls-break-up

**Reply by mhaberler** [reply]
it's pure speculation, but if I were to integrate observers into a UI editor: 

the UI editor probably will be agnostic about middleware - just too many of the latter

if so, a build artefact of UI creation could be a list of observer name/type tuples in the minimum

those then could be tranformed into a mapping table  `{topic, transformation, observername, observertype}`

if there'd be a way to associate structured userdata with an observer in the UI editor proper and have that exported to go along into the build step then `topic` and `transformation`  could be done in the UI editor as well

I guess we need to consult @kisvegabor how he thinks this is going to evolve

**Reply by sukesh-ak** [reply]
I have used `lv_observer` with `lvgl 9` in a recent project and it works well.
There are lot of features in it but not enough documented samples. I need to find time to do that.

**Reply by mhaberler** [reply]
same experience - pretty robust, no issues

I experimented with posting stringified JSON to a string observer, and do the [JSON parsing and extraction in the message callback](https://github.com/mhaberler/arduino3-playground/blob/lvgl8-observer/src/lv_custom_observers.cpp#L103-L289)

it works, but I do not like it - it is a bad separation of concerns, and it nails extraction to C++ code at least the way I did it

btw there's always the `user_data` field in LVGL objects; if one could populate that in the UI editor that would be a vehicle to communicate transformation/topic as well

**Reply by sukesh-ak** [reply]
Looks like you have gone to the next level with the Observers :)

**Reply by mhaberler** [reply]
well I thought observers which comprehend JSON directly would make a nice interface

doesnt matter if stuff comes via MQTT, HTTP, Serial, [NFC](https://github.com/mhaberler/arduino3-playground/blob/lvgl8-observer/src/nfcreader.cpp#L192-L202), whatever - or internally - dump a string onto observer, done

nice thing - observer names would map 1:1 to MQTT topics, it just shifts any transformation into to the UI

what do you think about that approach?

**Reply by sukesh-ak** [reply]
Sounds like a `data provider` which is mapped to Observer. I like the thought process.

**Reply by kisvegabor** [reply]
> LVGL & SquareLine has parted ways. So I am looking forward to what comes next from LVGL side
I guess we need to consult @kisvegabor how he thinks this is going to evolve

We have really parted at the beginning of this year. These are options I see now:
1. Ask around in [SquareLine's Forum](https://forum.squareline.io/) about their roadmap
2. Give a try to [EEZ Studio](https://github.com/eez-open/studio) (free and open-source, plus the developers are really nice and responsive)
3. Implement the observer part independently of the UI editor. I can share the v8 backported version of the observer if needed.
4. Not really an option, I mention it just for the sake of completeness: at LVGL we are also working on "something" but I cannot share the details yet.

**Reply by mhaberler** [reply]
https://forum.squareline.io/t/q-plans-timeline-for-lvgl-9-observer-support/2782

https://github.com/eez-open/studio/issues/311

**Reply by mhaberler** [reply]
the answers are in:
EEZ Studio: [LVGL9 now, observers out of scope](https://github.com/eez-open/studio/issues/311#issuecomment-2039375994)
Squareline:  [LVGL9 per April release,  obervers supported later](http://forum.squareline.io/t/q-plans-timeline-for-lvgl-9-observer-support/2782/2?u=haberlerm)

to leave the choice open, this would suggest not to assume observer support in the UI design tool - matching @kisvegabor 's third option

I did a lvgl8 backport of observers because currently using Squareline (8 only as of now), and it was straightforward - should be safe to go to lvgl9 and play with EEZ Studio meanwhile as Sqareline gets ready for 9

**Reply by sukesh-ak** [reply]
Squareline studio UI is pretty bad in design specially performance. Not sure what UI framework they are using.
EEZ Studio has pretty good in performance. Though I still like to handcraft UI sometimes so `lv_observer` is easier for me to include.

**Reply by mhaberler** [reply]
I think it's [Unity](https://en.wikipedia.org/wiki/Unity_(game_engine))

**Reply by kisvegabor** [reply]
Yes, it's Unity.

**Comment by mhaberler** [comment]
@sukesh-ak I've been playing with integrating [PicoMQTT](https://github.com/mlesniew/PicoMQTT) into my [playground](https://github.com/mhaberler/arduino3-playground/blob/lvgl8-observer/src/mqtt_ops.cpp)

found it to work nicely as a broker fed both with [OMG](https://github.com/1technophile/OpenMQTTGateway) and [esp32-ble2mqtt]( https://github.com/shmuelzon/esp32-ble2mqtt)

I think broker + UI is an interesting combination as it can get rid of an external broker in many cases

PicoMQTT is right now MQTT-over-TCP only but should sport Websockets later this spring - can then serve web MQTT clients + publishers as well

ok, it's an Arduino component and as such probably not to your liking, but it might serve as a blueprint for an ESP-IDF component

**Reply by sukesh-ak** [reply]
Should be able to port it to ESP-IDF without much work. Need to check the license. 

I also have a sample here on how to use Arduino library with ESP-IDF
https://github.com/sukesh-ak/Arduino-Library-with-IDF

**Reply by mhaberler** [reply]
I looked into that, and [one part it not quite compatible](https://github.com/mlesniew/PicoMQTT/blob/master/src/PicoMQTT/connection.cpp#L62-L78) with pure eventloop operation but I guess @mlesniew might have an opinion?

**Reply by mhaberler** [reply]
see https://github.com/mlesniew/PicoMQTT/issues/32https://github.com/mlesniew/PicoMQTT/issues/32

**Reply by mhaberler** [reply]
@mlesniew just solved the issue: https://github.com/mlesniew/PicoMQTT/issues/32#issuecomment-2094932017

**Comment by mhaberler** [comment]
meanwhile, on the embedded MQTT broker front: 

PicoMQTT now also supports [MQTT-over-Websockets](https://github.com/mlesniew/PicoMQTT/blob/websockets/examples/websocket_server/websocket_server.ino), opening the path to a web client with say MQTT.js or paho-mqtt

[working example ](https://github.com/mhaberler/picomqtt-tests)

**Reply by sukesh-ak** [reply]
Several MQTT clients are mentioned in my write-up here including `MQTT Web Client`. 
https://github.com/sukesh-ak/setup-mosquitto-with-docker

**Reply by mhaberler** [reply]
have a look at https://github.com/adrianromero/myhelloiot/issues/6#issuecomment-2080185332

(can you drop me a private mail or IM contact?)

**Reply by sukesh-ak** [reply]
My Github profile has my linked-in.
You can reach me => sukesh at yawsoftware.com

**Reply by mhaberler** [reply]
I think myhelloiot looks like a good set of functionality for generic MQTT dash board:  
- several stock widgets (LVGL has enough of those for a start)
- a way to assoicate a widget with a subscription
- a way to [extract a value from a structured message](https://github.com/adrianromero/myhelloiot/issues/6#issuecomment-2079862197)
- a way to set the broker (or just use an internal one with PicoMQTT)

anything else? I think that would be pretty useful already
